### PR TITLE
Apply localization to all hard coded text

### LINKED ISSUE #2408: Apply localization to all hard coded text in the Expensify.cash app
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/master/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!

Upwork job - https://www.upwork.com/jobs/~01ad599888651ad174
___

## Expected Result:
We are looking to apply localization to the Expensify.cash app, so that all hard coded text is replaced with locale-aware components/ methods. This project will require (1) Identifying all places that display hardcoded texts in the app, and (2) Using the `withLocalize` HOC in this PR https://github.com/Expensify/Expensify.cash/pull/2208 to make all those texts support localized versions.

## Actual Result:
Today, all text in the Expensify.cash app is hardcoded in English. Here are a few examples for reference:

image

image

## Platform:
This should apply to all platforms listed:

- [ ] Web
- [ ] iOS
- [ ] Android
- [ ] Desktop App
- [ ] Mobile Web

**Expensify/Expensify Issue URL:** https://github.com/Expensify/Expensify/issues/152153

## Other considerations:
There may be a few areas where it doesn't make sense to translate hardcoded text, which are important to highlight as a part of this project. For example, we should not update the hard corded text found here:

- Click avatar (above chat list) > Profile > Set my time zone manually > see dropdown list of regions
exemption

### PR DESCRIPTION
**Pull request by eVoloshchak**
### Details
Applied localization to the Expensify.cash app, all hard coded text is replaced with locale-aware components/ methods.

### Fixed Issues
Fixes https://github.com/Expensify/Expensify.cash/issues/2408

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
Leaving out the screenshots because there are no visual changes

### CODE REVIEW

**Review by pecanoro — DISMISSED** [review]

**Comment by pecanoro** [comment]
It looks good to me, but I would love a second opinion since the PR is massive and React Native is not my strongest suit. @roryabraham or @Julesssss could any of you take a look?

**Comment by Julesssss** [comment]
Sorry @pecanoro, but I'm unable to take on additional reviews just yet -- if this is still open once my N5 blockers are cleared I'll happily take a look though.

**Comment by roryabraham** [comment]
I can take a look!

File: src/CONST.js [hunk 1]
```diff
@@ -67,8 +67,8 @@ const CONST = {
     },
     MESSAGES: {
```
Comment (roryabraham) [inline src/CONST.js:68]: It's very odd to me that, out of all the strings in the app, just these two messages and the pronouns are included in this CONST file. I think that there are two ways we could handle this:

1. Move _all_ strings into a `UI_STRINGS.js` constant file, that looks sort of like this:

    ```javascript
    CANCEL: 'common.cancel',
    ...
    NO_PHONE_NUMBER: 'noPhoneNumberMessage',
    ...
    
    ```
    
    Where each constant provides a `lodashGet` path to a string in one of our language files.
    
2. Just remove these constants from `CONST.js` and use their `lodashGet` path directly in the component that's rendering them (like you have for most other components in this PR).

@iwiznia, do you have any opinion on which is better? I would lean towards (2) just because it's a bit simpler, but in Mobile-Expensify we had a `UIStrings` constants file.

File: src/components/withLocalize.js [hunk 2]
```diff
@@ -11,14 +11,12 @@ import numberFormat from '../libs/numberFormat';
...
     // Translations functions using current User's preferred locale
-    translations: PropTypes.shape({
-        translate: PropTypes.func.isRequired,
-        numberFormat: PropTypes.func.isRequired,
-        timestampToRelative: PropTypes.func.isRequired,
-        timestampToDateTime: PropTypes.func.isRequired,
-        toLocalPhone: PropTypes.func.isRequired,
-        fromLocalPhone: PropTypes.func.isRequired,
-    }),
+    translate: PropTypes.func.isRequired,
```
Comment (roryabraham) [inline src/components/withLocalize.js:14]: I know you didn't write this component, but each of these props should have a comment description

File: src/pages/home/report/EmojiPickerMenu/index.js [hunk 3]
```diff
@@ -333,7 +337,10 @@ class EmojiPickerMenu extends Component {
 EmojiPickerMenu.propTypes = propTypes;
 EmojiPickerMenu.defaultProps = defaultProps;
 
-export default withWindowDimensions(React.forwardRef((props, ref) => (
+export default compose(
+    withWindowDimensions,
+    withLocalize,
+)(React.forwardRef((props, ref) => (
```
Comment (roryabraham) [inline src/pages/home/report/EmojiPickerMenu/index.js:343]: Same problem here with `forwardRef`

File: src/pages/home/report/ReportTypingIndicator.js [hunk 4]
```diff
@@ -56,7 +59,7 @@ class ReportTypingIndicator extends React.Component {
                         ]}
                         >
                             <Text style={[styles.textStrong]}>{getDisplayName(this.state.usersTyping[0])}</Text>
-                            {' is typing...'}
+                            {this.props.translate('reportTypingIndicator.isTyping')}
```
Comment (roryabraham) [inline src/pages/home/report/ReportTypingIndicator.js:62]: Seems bad to include leading/trailing spaces in the translation. Instead, just translate the phrase, and include leading/trailing spaces in the component:

```suggestion
                            {` ${ this.props.translate('reportTypingIndicator.isTyping')}`}
```

File: src/pages/home/report/ReportTypingIndicator.js [hunk 5]
```diff
@@ -69,9 +72,9 @@ class ReportTypingIndicator extends React.Component {
                         ]}
                         >
                             <Text style={[styles.textStrong]}>{getDisplayName(this.state.usersTyping[0])}</Text>
-                            {' and '}
+                            {this.props.translate('reportTypingIndicator.and')}
```
Comment (roryabraham) [inline src/pages/home/report/ReportTypingIndicator.js:75]:
```suggestion
                            {` ${this.props.translate('common.and')} `}
```

**Comment by eVoloshchak** [comment]
> Sorry @pecanoro, but I'm unable to take on additional reviews just yet -- if this is still open once my N5 blockers are cleared I'll happily take a look though.

I think @iwiznia also [wanted to take a look ](https://github.com/Expensify/Expensify.cash/pull/2562#pullrequestreview-645123362)

File: src/pages/home/report/ReportTypingIndicator.js [hunk 6]
```diff
@@ -69,9 +72,9 @@ class ReportTypingIndicator extends React.Component {
...
                             <Text style={[styles.textStrong]}>{getDisplayName(this.state.usersTyping[1])}</Text>
-                            {' are typing...'}
+                            {this.props.translate('reportTypingIndicator.areTyping')}
```
Comment (roryabraham) [inline src/pages/home/report/ReportTypingIndicator.js:77]:
```suggestion
                            {` ${this.props.translate('reportTypingIndicator.areTyping')}`}
```

File: src/pages/home/report/ReportTypingIndicator.js [hunk 7]
```diff
@@ -83,8 +86,10 @@ class ReportTypingIndicator extends React.Component {
                             styles.chatItemComposeSecondaryRowOffset,
                         ]}
                         >
-                            <Text style={[styles.textStrong]}>Multiple users</Text>
-                            {' are typing...'}
+                            <Text style={[styles.textStrong]}>
+                                {this.props.translate('reportTypingIndicator.multipleUsers')}
+                            </Text>
+                            {this.props.translate('reportTypingIndicator.areTyping')}
```
Comment (roryabraham) [inline src/pages/home/report/ReportTypingIndicator.js:92]:
```suggestion
                            {` ${this.props.translate('reportTypingIndicator.areTyping')`}
```

File: src/languages/en.js [hunk 8]
```diff
@@ -1,4 +1,231 @@
...
+        copyLink: 'Copy Link',
+        markAsUnread: 'Mark as Unread',
+        editComment: 'Edit Comment',
+        deleteComment: 'Delete Comment',
+    },
+    reportActionsView: {
+        beFirstPersonToComment: 'Be the first person to comment',
+    },
+    reportTypingIndicator: {
+        isTyping: ' is typing...',
```
Comment (roryabraham) [inline src/languages/en.js:85]:
```suggestion
        isTyping: 'is typing...',
```

File: src/languages/en.js [hunk 9]
```diff
@@ -1,4 +1,231 @@
...
+        and: ' and ',
```
Comment (roryabraham) [inline src/languages/en.js:86]:
```suggestion
```

File: src/languages/en.js [hunk 10]
```diff
@@ -1,4 +1,231 @@
...
+        areTyping: ' are typing...',
```
Comment (roryabraham) [inline src/languages/en.js:87]:
```suggestion
        areTyping: 'are typing...',
```

File: src/pages/home/report/ReportTypingIndicator.js (hunk 5 above)
Comment (roryabraham) [reply src/pages/home/report/ReportTypingIndicator.js:75]: Also, I think `and` should be in the `common` translations section.

**Review by roryabraham — CHANGES_REQUESTED** [review]

File: src/components/TextInputFocusable/index.js [hunk 11]
```diff
@@ -235,7 +238,7 @@ class TextInputFocusable extends React.Component {
 TextInputFocusable.propTypes = propTypes;
 TextInputFocusable.defaultProps = defaultProps;
 
-export default React.forwardRef((props, ref) => (
+export default withLocalize(React.forwardRef((props, ref) => (
```
[the remark this answers is not in the document]
Comment (roryabraham) [reply src/components/TextInputFocusable/index.js:241]: You can look at `withWindowDimensions` for an example of how to make a HOC forward a ref. Basically it has to use `React.forwardRef` to take the ref in as a prop, and then pass that prop as a ref to the wrapped component.

**Review by iwiznia — CHANGES_REQUESTED** [review]
Nice! Just a few comments

**Comment by eVoloshchak** [comment]
Updated

File: src/components/withLocalize.js [hunk 12]
```diff
@@ -10,17 +10,27 @@ import {toLocalPhone, fromLocalPhone} from '../libs/LocalePhoneNumber';
...
+    // Returns a locally converted phone number without the country code
     toLocalPhone: PropTypes.func.isRequired,
+
+    // Returns an internationally converted phone number with the country code
     fromLocalPhone: PropTypes.func.isRequired,
 };
 
 function withLocalizeHOC(WrappedComponent) {
-    const withLocalize = (props) => {
+    const WithLocalize = (props) => {
```
Comment (iwiznia) [inline src/components/withLocalize.js:33]: Why this change? I think variables should not start with capital per our style guide, no?

File: src/languages/en.js [hunk 13]
```diff
@@ -9,7 +9,7 @@ export default {
         to: 'To',
         optional: 'Optional',
         split: 'Split',
-        request: 'Request',
+        request: ({amount}) => `Request ${amount}`,
```
Comment (iwiznia) [inline src/languages/en.js:12]: Given that this is `Request ${amount}` should we move this outside `common`?

**Review by iwiznia — CHANGES_REQUESTED** [review]

File: src/languages/en.js (hunk 13 above)
Comment (eVoloshchak) [reply src/languages/en.js]: > Given that this is `Request ${amount}` should we move this outside `common`?

Moved to 'iou'

**Review by iwiznia — APPROVED** [review]

**Review by roryabraham — APPROVED** [review]

File: src/components/AttachmentPicker/index.native.js [hunk 14]
```diff
@@ -142,18 +71,18 @@ class AttachmentPicker extends Component {
...
-                pickAttachment: () => showImagePicker(RNImagePicker.launchImageLibrary),
+                text: this.props.translate('attachmentPicker.chooseFromGallery'),
+                pickAttachment: () => this.showImagePicker(RNImagePicker.launchImageLibrary),
             },
             {
                 icon: Paperclip,
-                text: 'Choose Document',
-                pickAttachment: showDocumentPicker,
+                text: this.props.translate('attachmentPicker.chooseDocument'),
+                pickAttachment: this.showDocumentPicker,
```
Comment (kidroca) [inline src/components/AttachmentPicker/index.native.js:85]: `this` reference is lost here, even if `showDocumentPicker` is bound like

```js
 this.showDocumentPicker =  this.showDocumentPicker.bind(this)
```

Because `bind` happens at the end of the `constructor` and the assigned `pickAttachment` here would be the `showDocumentPicker` before `bind` runs

If the `if` condition in `showDocumentPicker` ever evaluates to true the app would silently log an error, and the user will not be notified that attachment picking failed (Though I don't know when that would happen, no storage permissions?)
image


The easiest fix would be to change this to:

```js
pickAttachment: () => this.showDocumentPicker(),
```

File: src/components/AttachmentPicker/index.native.js [hunk 15]
```diff
@@ -174,6 +103,80 @@ class AttachmentPicker extends Component {
...
+        Alert.alert(
+            this.props.translate('attachmentPicker.cameraPermissionRequired'),
+            this.props.translate('attachmentPicker.expensifyDoesntHaveAccessToCamera'),
+            [
+                {
+                    text: this.props.translate('common.cancel'),
+                    style: 'cancel',
+                },
+                {
+                    text: this.props.translate('common.settings'),
```
Comment (kidroca) [inline src/components/AttachmentPicker/index.native.js:119]: No such key `common.settings` in languages/en: https://github.com/Expensify/Expensify.cash/pull/2718/files#diff-22037c1e4dbdea8803b450b3fce06e6c475762a31ecf84b34dc264ba0e443637R3

If it ever becomes the case that the user revoked camera permissions the app will seem to do nothing when they press on "Add Attachment" -> "Take Photo". 
While the app actually throws an error here 

https://user-images.githubusercontent.com/12156624/121061991-38cb4200-c7cd-11eb-8d38-6be3e5921773.mp4

The problem will be worse if the user selects "Deny & don't ask" as nothing would seem to happen after you press on "Take Photo"

- Expected: When the app lacks permissions to take photos for some reason it should show alert message explaining why the "Take Photo" button doesn't work 
- Actual: Pressing "Take Photo" does nothing

**Comment by iwiznia** [comment]
@kidroca great catches! @eVoloshchak can you send a PR fixing them please?

File: src/languages/en.js [hunk 16]
```diff
@@ -1,4 +1,233 @@
...
+        phoneNumber: 'Phone Number',
+        email: 'Email',
+        and: 'and',
+    },
+    attachmentPicker: {
+        cameraPermissionRequired: 'Camera Permission Required',
+        expensifyDoesntHaveAccessToCamera: 'Expensify.cash does not have access to your camera, please enable the permission and try again.',
+        attachmentError: 'Attachment Error',
+        errorWhileSelectingAttachment: 'An error occurred while selecting an attachment, please try again',
+        errorDuringAttachmentSelection: 'Error during attachment selection',
```
Comment (kidroca) [inline src/languages/en.js:35]: aren't `errorWhileSelectingAttachment` and `errorDuringAttachmentSelection` pretty much the same thing said differently

**Comment by eVoloshchak** [comment]
> @kidroca great catches! @eVoloshchak can you send a PR fixing them please?

Submitted a [PR](https://github.com/Expensify/Expensify.cash/pull/3435)
